akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala

```diff
@@ -0,0 +1,135 @@
+
+  val timeout = 300.milliseconds
+  val bytes = "abc".getBytes
+  val byteString = ByteString("abc")
+
+  def expectBlocked[T](f: Future[T]) =
+    the[Exception] thrownBy Await.result(f, timeout) shouldBe a[TimeoutException]
+
+  def expectSuccess[T](f: Future[T], value: T) =
+    Await.result(f, timeout) should be(value)
+
+  "OutputStreamSource" must {
+    "read bytes from OutputStream" in assertAllStagesStopped {
+      val (outputStream, probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      outputStream.write(bytes)
+      s.request(1)
+      probe.expectNext(byteString)
+      outputStream.close()
+      probe.expectComplete()
+      //Await.result(f, timeout) should be(3)
+    }
+
+    "block flush call until send all buffer to downstream" in assertAllStagesStopped {
+      val (outputStream, probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      outputStream.write(bytes)
+      val f = Future(outputStream.flush())
+
+      expectBlocked(f)
+      probe.expectNoMsg(Zero)
+
+      s.request(1)
+      expectSuccess(f, ())
+      probe.expectNext(byteString)
+
+      outputStream.close()
+      probe.expectComplete()
+    }
+
+    "not block flushes when buffer is empty" in assertAllStagesStopped {
+      val (outputStream, probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      outputStream.write(bytes)
+
+      val f = Future(outputStream.flush())
+      s.request(1)
+      expectSuccess(f, ())
+      probe.expectNext(byteString)
+
+      val f2 = Future(outputStream.flush())
+      expectSuccess(f2, ())
+
+      outputStream.close()
+      probe.expectComplete()
+    }
+
+    "block writes when buffer is full" in assertAllStagesStopped {
+      val (outputStream, probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      //default max-input-buffer-size = 16
+      (1 to 16).foreach { _ ⇒ outputStream.write(bytes) }
+
+      //blocked call
+      val f = Future(outputStream.write(bytes))
+
+      expectBlocked(f)
+      probe.expectNoMsg(Zero)
+
+      s.request(17)
+      expectSuccess(f, ())
+      probe.expectNextN(List.fill(17)(byteString).toSeq)
+
+      outputStream.close()
+      probe.expectComplete()
+    }
+
+    "throw error when write after stream is closed" in assertAllStagesStopped {
+      val (outputStream, probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+
+      probe.expectSubscription()
+      outputStream.close()
+      probe.expectComplete()
+      the[Exception] thrownBy outputStream.write(bytes) shouldBe a[IOException]
+    }
+
+    "throw IOException when writing to the stream after the subscriber has cancelled the reactive stream" in assertAllStagesStopped {
+      pending
```

Comment:
This test case requires some synchronization. We need to know when stage received "canceled" message from downstream. I created test version of SourceState for previous implementation - will figure something out with GraphStage as well. Until the whole approach is not approved would like to keep it pending